src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -837,21 +837,25 @@ private void FreeOverlapped(bool checkForShutdown)
                 if (_socketAddressGCHandle.IsAllocated)
                 {
                     _socketAddressGCHandle.Free();
+                    _pinnedSocketAddress = null;
```

Comment:
That check will still function in the case where the buffer was not unpinned before the SocketAsyncEventArgs object was reused.  E.g., if you're reusing the same buffer *and* address for a subsequent request.

The problem was that if you free'd the buffer, then made a new request with a new buffer but the *same* address, then the address would compare equal to the old one, and we'd assume it was already pinned - but of course it was un-pinned along with the buffer.